### DISCUSSION TITLE
有没有考虑过把官方维护的插件发布为Nuget包

### DISCUSSION THREAD

**Discussion by Alex1911-Jiang** [yiyungent/KnifeHub#23] [General]
我的插件需要依赖 KonataPlugin ，但是如果这个插件更新，我要拉取源代码再编译使用起来不太方便，如果能通过 Nuget 就好了🙏

**Comment by yiyungent** [comment]
可以是可以，不过感觉有点滥用 nuget（下次有时间我看情况，传下吧），目前建议通过 git submodule 方式来开发插件
（其实你自己插件需要用到的部分对应上 KonataPlugin 没变, 那么插件还是可以正常加载，不会因为 KonataPlugin 插件一更新, 你的插件就没法用了）

这里提醒一下，目前还不支持直接上传 nupkg 包插件（那个上传页面提示支持，但是需要将所有用到的dll一起打包进入nupkg，而默认 dotnet pack 是不会这么做的），支持直接上传 nupkg 插件目前在 PluginCore 已经搁置了，解析依赖再下载再解析，已经发现太多问题了，代码我也还没传上去（逃

后续 `PluginStore` 将搜索 GitHub 仓库所有拥有特定标识为插件的仓库，利用 Releases 中的 Assets 来当插件来源 (APlugin-v0.1.0.zip)，`APlugin.nupkg` 将被废弃

**Comment by Alex1911-Jiang** [comment]
好吧，主要是我小白没太整明白 submodule 怎么用🤡
我暂时先引用着项目开发吧
顺带一提使用插件模板创建项目的时候如果项目名字有小数点会有亿点点问题

**Reply by yiyungent** [reply]
> 好吧，主要是我小白没太整明白 submodule 怎么用🤡 我暂时先引用着项目开发吧 顺带一提使用插件模板创建项目的时候如果项目名字有小数点会有亿点点问题

dalao 谦虚了，下次我就把插件传上 nuget, 
后面我看看插件模板，这玩意还是建议全英文，毕竟小数点也是命名空间分割符，
建议开发时参考一下 其它依赖了 KonataPlugin 的项目.csproj， 打包插件这步很容易出问题

**Comment by yiyungent** [comment]
@Alex1911-Jiang 
好了，不知道发布后可行不，记得要在打包时排除，目前只发布了 KonataPlugin 的，可用的话再发布其它插件
https://www.nuget.org/packages/PluginCore.KonataPlugin

```xml
<ItemGroup>
    <PackageReference Include="PluginCore.KonataPlugin" Version="0.1.10">
	    <ExcludeAssets>runtime</ExcludeAssets>
    </PackageReference>
</ItemGroup>
```

**Comment by Alex1911-Jiang** [comment]
emmmm，好像不是很行，引用nuget包之后编译的QQHelloWorldPlugin，上传和安装都能成功，但是启用的时候报
启用失败: One or more errors occurred. (Unable to load one or more of the requested types. )
然后我以为是我之前装的 KonataPlugin （在Release下载的）导致的，然后我把它卸了重装，结果它点启用的时候也报这个错了

KnifeHub 版本 v1.2.1

编译的时候有添加排除
```XML
	<ItemGroup>
		<PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0">
			<ExcludeAssets>runtime</ExcludeAssets>
		</PackageReference>
    		<PackageReference Include="PluginCore.KonataPlugin" Version="0.1.10">
	    		<ExcludeAssets>runtime</ExcludeAssets>
    		</PackageReference>
	</ItemGroup>
```
还有就是没引用 nuget 包，直接拉取代码什么都没改编译的 QQHelloWorldPlugin 在上传的时候会报空对象异常：
![I{ ZCF@DL)U I35%G3QQYUI](https://user-images.githubusercontent.com/50268952/218301730-9460af03-7606-44a5-be27-ca139b7f3c2c.png)
Release上的能正常安装使用，但我对比了一下我编译的只少一个Ref文件夹

**Comment by Alex1911-Jiang** [comment]
似乎确实是我编译的插件会导致问题，我把 KnifeHub 删了重新解压，KonataPlugin 是能成功安装并启用的，但是尝试启用我编译的 QQHelloWorldPlugin 之后，别的插件也无法启用，并且原来正常安装和启用的 KonataPlugin 在禁用后点击卸载也会报插件不存在的错误，无法卸载

**Reply by yiyungent** [reply]
> 似乎确实是我编译的插件会导致问题，我把 KnifeHub 删了重新解压，KonataPlugin 是能成功安装并启用的，但是尝试启用我编译的 QQHelloWorldPlugin 之后，别的插件也无法启用，并且原来正常安装和启用的 KonataPlugin 在禁用后点击卸载也会报插件不存在的错误，无法卸载

???这么奇怪，注意观察一下 `App_Data/plugin.config.json` 与 `Plugins` 文件夹，插件状态不正常与这些有关，

最新版本的 [KnifeHub-v1.2.1](https://github.com/yiyungent/KnifeHub/releases/tag/KnifeHub-v1.2.1) 已经支持如下接口，可以访问下看看有没有啥缺失

*   `api/PluginCore/Admin/Debug/PluginContexts`
*   `api/PluginCore/Admin/Debug/AssemblyLoadContexts`
*   `api/PluginCore/Admin/Debug/Assemblies`
*   `api/PluginCore/Admin/Debug/Services`

**Reply by Alex1911-Jiang** [reply]
api/PluginCore/Admin/Debug/Assemblies 这个不太正常
```Json
{
"code": -1,
"message": "error",
"data": "System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.\r\n\r\n at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)\r\n at System.Reflection.RuntimeAssembly.get_DefinedTypes()\r\n at PluginCore.AspNetCore.Controllers.DebugController.Assemblies()\r\nSystem.IO.FileNotFoundException: \r\nFile name: 'KonataPlugin, Version=0.1.10.0, Culture=neutral, PublicKeyToken=null'"
}
```

**Comment by yiyungent** [comment]
`One or more errors occurred. (Unable to load one or more of the requested types. )`

可能是啥重复加载了，还有有时候依赖顺序不正确也可能导致（先加载 KonataPlugin，再加载 依赖插件，建议启用 KonataPlugin 后，随便填下账号密码，点击登录，确保 KonataPlugin 其中一些dll先加载，加载采用被动懒加载，在没有碰到类型前，不会主动加载非APlugin.dll）

**Comment by yiyungent** [comment]
@Alex1911-Jiang 

`System.IO.FileNotFoundException: \r\nFile name: 'KonataPlugin, Version=0.1.10.0, Culture=neutral, PublicKeyToken=null`

注意看下最后一句，是不是 `Plugins` 下缺失了，QQHelloWordPlugin 依赖 KonataPlugin, 那么 QQHelloWordPlugin  下是没有KonataPlugin.dll 的（以及 KonataPlugin已经包含的dll包） ，我一会再试试，并把仓库内依赖KonataPlugin的插件Release改为nuget引用，

还真是，我试了下把 QQHelloWorldPlugin 改为nuget ，
![image](https://user-images.githubusercontent.com/16939388/218388054-44d065c5-5c1b-410f-9bf0-1b9377cb8c63.png)

**Comment by yiyungent** [comment]
@找到问题所在了，因为签名没变，KonataPlugin 匹配不上，我发布个新版本 KonataPlugin  指定一下版本
![image](https://user-images.githubusercontent.com/16939388/218389140-393c912f-898a-4099-8287-dd27f4da96d7.png)

![image](https://user-images.githubusercontent.com/16939388/218389122-665fc1cd-5d8e-4a54-88b3-37f9437716cf.png)

**Comment by yiyungent** [comment]
**突然想起来，为啥不发nuget包了**，

因为这样会导致你用引用nuget包发布的插件，最终引用依赖项（插件.dll的依赖信息会带上nuget版本号），这样在搜索时，就会必须匹配版本号才行，这样插件依赖就不能兼容了，A插件更新版本后，B插件也必须更新

> 最新 KonataPlugin 我试了固定 `<AssemblyVersion>1.0.0.0</AssemblyVersion>`

不过建议还是用项目依赖吧，配合上 git submodule，最近挺忙，nuget包方式怪麻烦的
